libcudacxx/include/cuda/pipeline

```diff
@@ -514,59 +516,45 @@ _LIBCUDACXX_BEGIN_NAMESPACE_CUDA
     {
         (void)__pipeline;
         NV_IF_TARGET(NV_PROVIDES_SM_80,
-            __memcpy_arrive_on_impl::__arrive_on(__barrier, async_contract_fulfillment::async);
+            __barrier_cp_async_arrive_on(__barrier);
         )
     }
 
-    template<typename _Group, class _Type, thread_scope _Scope>
-    _LIBCUDACXX_INLINE_VISIBILITY
-    async_contract_fulfillment memcpy_async(_Group const & __group, _Type * __destination, _Type const * __source, std::size_t __size, pipeline<_Scope> & __pipeline) {
-        // When compiling with NVCC and GCC 4.8, certain user defined types that _are_ trivially copyable are
-        // incorrectly classified as not trivially copyable. Remove this assertion to allow for their usage with
-        // memcpy_async when compiling with GCC 4.8.
-        // FIXME: remove the #if once GCC 4.8 is no longer supported.
-    #if !defined(_LIBCUDACXX_COMPILER_GCC) || _GNUC_VER > 408
-        static_assert(std::is_trivially_copyable<_Type>::value, "memcpy_async requires a trivially copyable type");
-    #endif
-
-        return __memcpy_async<alignof(_Type)>(__group, reinterpret_cast<char *>(__destination), reinterpret_cast<char const *>(__source), __size, __pipeline);
-    }
+    template<thread_scope _Sco, bool _Tx, typename _Arch, __space _OutSpace, __space _InSpace, __space _SyncSpace>
+    struct __memcpy_async_sync_hooks<
+        pipeline<_Sco>, _Tx, _Arch,
+        _OutSpace, _InSpace, _SyncSpace
```

Comment:
For sync_hooks specifically, and for _Arch specifically, maybe? I can experiment with that. _Sco is an inherent part of the pipeline type, so no on that one. Also no on the same question applied to the main hooks, because there the arch being matched early is fundamental to how I'm eliminating *a lot* of template instantiations and generating code (before I did that, I'd have ptxas take upwards of 50 gigs of memory trying to make sense out of the thing I was spewing out).

Reply:
I think this would make it annoying to do the "fallback" for barrier, because the fallback happens for both non-sm80+ arches, but also for non-shared spaces; as currently spelled out, sure, there's a bit of sfinae, but then we explicitly inherit from `__noop_sync_traits`, which I think is a self documenting piece of code. Without the arch in the struct's template parameter list, we'd now kinda have two places that do that fallback, and the global->shared spec would also need to inherit from `__noop_sync_traits`, which I think is awkward.

Reply:
Sounds good